File: scilab/modules/differential_equations/sci_gateway/fortran/Ex-daskr.f
c Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
c Copyright (C) INRIA
c ...
c 
c This file must be used under the terms of the CeCILL.
c This source file is licensed as described in the file COPYING, which
c you should have received as part of this distribution.  The terms
c are also available at    
c http://www.cecill.info/licences/Licence_CeCILL_V2-en.txt
c
c     Examples for daskr
c     ----------------------
c
c
      subroutine pjac1(res, ires, neq, t, y, ydot, rewt, savr,
     *                   wk, h, cj, wp, iwp, ier, rpar, ipar)
      double precision t, y(*), ydot(*), rewt(*), savr(*),
     *                  wk(*), h, cj, wp(*), rpar(*)
      double precision ysave, ypsave, e(neq), del, delinv
      external res
      integer ires, neq, iwp(*), ier, ipar(*)
      ier = -1
      uround = D1MACH (4)
c     Compute the preconditioner P = dG/dy + cj*dG/dydot, LU-decompose it
c and stock it in wp and iwp
      nrow = 0
      ires = 0
      SQuround = sqrt(uround)
      do 1 i=1,neq
	 del = max (SQuround*max(abs(y(i)),abs(h*ydot(i))), 1.0D0/rewt(i))
	 del = sign(del, h*ydot(i))
	 ysave   =  y(i)
	 ypsave  =  ydot(i)
	 y(i)    = y(i) + del
	 ydot(i) = ydot(i) + cj*del
	 call res(t,y,ydot,e,ires,rpar,ipar)
      if (ires.lt.0) return
	 delinv = 1.0D0/del
	 do 2 j=1,neq
		 wp(nrow+j) = (e(j)-savr(j))*delinv
2      continue
	 nrow = nrow + neq
	 y(i)    =  ysave
	 ydot(i) =  ypsave
	 iwp(i) = i
	 iwp(i+neq) = i
1      continue
c     Do dense-matrix LU decomposition on P
      lipvt = iwp(30)
      call dgefa(wp, neq , neq, iwp(lipvt), ier)
      return
      end

      subroutine psol1(neq, t, y, ydot, savr, wk, cj, wght,
     *                   wp, iwp, b, eplin, ier, rpar, ipar)
      double precision t, y(*), ydot(*), savr(*), wk(*), cj, wght(*),
     *                  wp(*), b(*), eplin, rpar(*)
      integer neq, iwp(*), ier, ipar(*)
      integer ipiv(neq)
      nColB = 1
      info = 0
c     b = P\b
c     Call DGESV to solve the system Px = b, where P is stored in
c wp in compact row format
      call dgesv (neq, nColB, wp, neq, ipiv, b, neq, info)
      if (info.ne.0) then
          ier = -1
          return
      endif
      ier = 0
      return
      end

File: scilab/modules/differential_equations/sci_gateway/c/Ex-daskr.c
#include <stdio.h>
#include <stdlib.h>
#include "machine.h"
#include "core_math.h"

extern void   C2F(dgefa) (double *A, int *lead_dim_A, int *n, int *ipivots, int *info);
extern void   C2F(dgesl) (double *A, int *lead_dim_A, int *n, int *ipivots, double *B, int *job);
extern double C2F(dlamch) (const char*);
typedef void (*resfunc) (double*, double*, double*, double*, int*, double*, int*);

void C2F(pjac1) (resfunc res, int *ires, int *neq, double *tOld, double *actual, double *actualP,
                 double *rewt, double *savr, double *wk, double *h, double *cj, double *wp, int *iwp,
                 int *ier, double *rpar, int *ipar);
void C2F(psol1) (int *neq, double *tOld, double *actual, double *actualP,
                 double *savr, double *wk, double *cj, double *wght, double *wp,
                 int *iwp, double *b, double *eplin, int *ier, double *dummy1, int *dummy2);

void C2F(pjac1) (resfunc res, int *ires, int *neq, double *tOld, double *actual, double *actualP,
                 double *rewt, double *savr, double *wk, double *h, double *cj, double *wp, int *iwp,
                 int *ier, double *rpar, int *ipar)
{
    int i = 0;
    int j = 0;
    int info = 0;
    int nrow = 0;
    double tx = 0;
    double del = 0;
    double delinv = 0;
    double ysave = 0;
    double ypsave = 0;
    double * e = NULL;

    double SQuround = sqrt(C2F(dlamch)("P"));

    tx = *tOld;

    e = (double *) calloc(*neq, sizeof(double));

    for (i = 0 ; i < *neq ; ++i)
    {
        del =  Max(SQuround * Max(fabs(actual[i]), fabs(*h * actualP[i])), 1. / rewt[i]);
        del *= (*h * actualP[i] >= 0) ? 1 : -1;
        del =  (actual[i] + del) - actual[i];
        ysave  = actual[i];
        ypsave = actualP[i];
        actual[i]  += del;
        actualP[i] += *cj * del;
        res(&tx, actual, actualP, e, ires, rpar, ipar);

        if (*ires < 0)
        {
            *ier = -1;
            free(e);
            return;
        }

        delinv = 1. / del;
        for (j = 0 ; j < *neq ; j++)
        {
            wp[nrow + j] = (e[j] - savr[j]) * delinv;
            /* NaN test */
            if (ISNAN(wp[nrow + j]))
            {
                *ier = -1;
                free(e);
                return;
            }
        }
        nrow       += *neq;
        actual[i]  =  ysave;
        actualP[i] =  ypsave;
    }

    /* Proceed to LU factorization of P. */
    C2F(dgefa) (wp, neq, neq, iwp, &info);
    if (info != 0)
    {
        *ier = -1;
    }

    free(e);
}

void C2F(psol1) (int *neq, double *tOld, double *actual, double *actualP,
                 double *savr, double *wk, double *cj, double *wght, double *wp,
                 int *iwp, double *b, double *eplin, int *ier, double *dummy1, int *dummy2)
{
    int i = 0, job = 0;

    C2F(dgesl) (wp, neq, neq, iwp, b, &job);

    /* NaN test */
    for (i = 0; i < *neq; ++i)
    {
        if (ISNAN(b[i]))
        {
            /* Indicate a recoverable error, meaning that the step will be retried with the same step size,
               but with a call to 'jacpsol' to update necessary data, unless the Jacobian data is current,
               if (b[i] - b[i] != 0) in which case the step will be retried with a smaller step size. */
            *ier = -1;
            return;
        }
    }
}
